reportssql/report_util_functions.sql

```diff
@@ -1483,4 +1483,223 @@ BEGIN
 
     RETURN result;
 END$$
-DELIMITER ;
\ No newline at end of file
+DELIMITER ;
+
+-- patientDiagnosedHIVPositiveBeforeReportEndDate
+
+DROP FUNCTION IF EXISTS patientDiagnosedHIVPositiveBeforeReportEndDate;
+
+DELIMITER $$
+CREATE FUNCTION patientDiagnosedHIVPositiveBeforeReportEndDate(
+    p_patientId INT(11),
+    p_endDate DATE) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+    DECLARE result TINYINT(1) DEFAULT 0;
+    DECLARE uuidHIVTestResult VARCHAR(38) DEFAULT "85dadffe-5714-4210-8632-6fb51ef593b6";
+    DECLARE uuidHIVTestResultPositive VARCHAR(38) DEFAULT "7acfafa4-f19b-485e-97a7-c9e002dbe37a";
+
+    SELECT
+        TRUE INTO result
+    FROM obs o
+    JOIN concept c ON c.concept_id = o.concept_id AND c.retired = 0
+    WHERE o.voided = 0
+        AND o.person_id = p_patientId
+        AND c.uuid = uuidHIVTestResult
+        AND o.value_coded IS NOT NULL
+        AND o.value_coded = (SELECT concept_id FROM concept WHERE uuid = uuidHIVTestResultPositive)
+        AND cast(o.obs_datetime AS DATE) < p_endDate
+    LIMIT 1;
+
+    RETURN (result );
+END$$
+DELIMITER ;
+
+-- patientHadANCVisitWithinReportPeriod
+
+DROP FUNCTION IF EXISTS patientHadANCVisitWithinReportPeriod;
+
+DELIMITER $$
+CREATE FUNCTION patientHadANCVisitWithinReportPeriod(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+    DECLARE result TINYINT(1) DEFAULT 0;
+    DECLARE uuidvisitTypeANC VARCHAR(38) DEFAULT "a71de1a2-aa43-496a-a533-13f47fad0129";
+
+    SELECT
+        TRUE INTO result
+    FROM visit v
+    JOIN visit_type vt ON vt.visit_type_id = v.visit_type_id AND vt.retired = 0
+    WHERE v.voided = 0
+        AND v.patient_id = p_patientId
+        AND vt.uuid = uuidvisitTypeANC
+        AND v.date_started IS NOT NULL
+        AND DATE(v.date_started) BETWEEN p_startDate AND p_endDate;
+
+    RETURN (result );
+END$$
+DELIMITER ;
+
+-- patientHIVRetestPositiveWithinReportingPeriod
+
+DROP FUNCTION IF EXISTS patientHIVRetestPositiveWithinReportingPeriod;
+
+DELIMITER $$
+CREATE FUNCTION patientHIVRetestPositiveWithinReportingPeriod(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+    DECLARE patientRetestedForHIVWIthinReportingPeriod TINYINT(1) DEFAULT 0;
+    DECLARE patientHIVRetestResultIsPositive TINYINT(1) DEFAULT 0;
+    DECLARE patientRepeatTestDateIsBetweenReportingPeriod TINYINT(1) DEFAULT 0;
+    DECLARE uuidRepeatTestDone VARCHAR(38) DEFAULT "4026c752-ecfe-43b0-9cf9-14a74d961fd1";
```

Comment:
@prashantsable1991 I thought you would remove these?

uuidRepeatTestDone
uuidRepeatTestIfYes

Reply:
Also, this function should consider the obs_group_id which should be "Prior to ANC Enrolment". Otherwise observations for "At ANC Enrolment" could be counted instead.